### PR TITLE
Barretenberg FFI  rust wrapper

### PR DESCRIPTION
**Pull request by guipublic**
Add dsl folder for noir
Add FFI rust wrapper for barretenberg

### CODE REVIEW

File: barretenberg_wrapper/src/bindings/bindings_hpp/blake2.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include <stdint.h>
+#include <stddef.h>
+//#include <vector>
```
Comment (kevaundray) [inline barretenberg_wrapper/src/bindings/bindings_hpp/blake2.hpp]: unused comment

File: barretenberg_wrapper/src/bindings/bindings_hpp/pedersen.hpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+//#include "pedersen.hpp"
```
Comment (kevaundray) [inline barretenberg_wrapper/src/bindings/bindings_hpp/pedersen.hpp]: These unused comments might have been mine. Can you remove it for me?

File: barretenberg_wrapper/src/blake2s.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+extern crate hex;
+
+//use crate::bindings;
```
Comment (kevaundray) [inline barretenberg_wrapper/src/blake2s.rs]: unused comment

File: barretenberg_wrapper/src/blake2s.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+use crate::bindings::blake2s;
+
+pub fn hash_to_field(/*&mut self,*/ input: &[u8]) -> [u8; 32] {
```
Comment (kevaundray) [inline barretenberg_wrapper/src/blake2s.rs]: unused &mut self

File: barretenberg_wrapper/src/composer.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        cs_ptr.as_ptr() as *const u8,
+        output_buf,
+    )
+}
+
+/// # Safety
+/// cs_prt must point to a valid constraints system structure of type standard_format
+pub unsafe fn get_circuit_size(cs_prt: *const u8) -> u32 {
+    dsl_standard_format::composer__get_circuit_size(cs_prt)
+    // TODO test with a circuit of size 2^19 cf: https://github.com/noir-lang/noir/issues/12
```
Comment (kevaundray) [inline barretenberg_wrapper/src/composer.rs:23]: We can do this TODO in a separate PR

File: barretenberg_wrapper/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+extern crate hex;
+
+mod bindings;
+pub mod blake2s;
+pub mod composer;
+pub mod pedersen;
+pub mod pippenger;
+pub mod schnorr;
+
+//pub use blake2s::hash_to_field;
```
Comment (kevaundray) [inline barretenberg_wrapper/src/lib.rs]: unused comment

File: barretenberg_wrapper/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+#[macro_use]
+#[cfg(test)]
+mod tests {
+    //    use super::*;
```
Comment (kevaundray) [inline barretenberg_wrapper/src/lib.rs]: There's nothing in this test module, lets remove it and add it back if we want to add tests

File: barretenberg_wrapper/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+//extern crate barretenberg_wrapper;
```
Comment (kevaundray) [inline barretenberg_wrapper/src/main.rs]: unused comment

File: barretenberg_wrapper/src/pedersen.rs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                expected_hex: "2619a3512420b4d3c72e43fdadff5f5a3ec1b0e7d75cd1482159a7e21f6c6d6a",
+            },
+        ];
+
+        for test in tests {
+            let got = compress_native(&test.input_left, &test.input_right);
+            let mut many_intputs: Vec<[u8; 32]> = Vec::new();
+            many_intputs.push(test.input_left);
+            many_intputs.push(test.input_right);
+            //test.input_left.to_vec().extend(test.input_right);
```
Comment (kevaundray) [inline barretenberg_wrapper/src/pedersen.rs]: unused comment, also on the line below

**Comment by kevaundray** [comment]
Looks good modulo the nits :)

File: barretenberg_wrapper/src/bindings/bindings_hpp/blake2.hpp (hunk 1 above)
Comment (guipublic) [reply barretenberg_wrapper/src/bindings/bindings_hpp/blake2.hpp]: done

File: barretenberg_wrapper/src/blake2s.rs (hunk 3 above)
Comment (guipublic) [reply barretenberg_wrapper/src/blake2s.rs]: done

File: barretenberg_wrapper/src/lib.rs (hunk 6 above)
Comment (guipublic) [reply barretenberg_wrapper/src/lib.rs]: done

File: barretenberg_wrapper/src/lib.rs (hunk 7 above)
Comment (guipublic) [reply barretenberg_wrapper/src/lib.rs]: done

File: barretenberg_wrapper/src/main.rs (hunk 8 above)
Comment (guipublic) [reply barretenberg_wrapper/src/main.rs]: done

File: barretenberg_wrapper/src/pedersen.rs (hunk 9 above)
Comment (guipublic) [reply barretenberg_wrapper/src/pedersen.rs]: done

File: barretenberg_wrapper/src/blake2s.rs (hunk 4 above)
Comment (guipublic) [reply barretenberg_wrapper/src/blake2s.rs]: done

File: barretenberg_wrapper/src/bindings/bindings_hpp/pedersen.hpp (hunk 2 above)
Comment (guipublic) [reply barretenberg_wrapper/src/bindings/bindings_hpp/pedersen.hpp]: done

**Review by kevaundray — APPROVED** [review]
